### PR TITLE
toolchain/binutils: add support for version 2.40

### PR DESCRIPTION
**Pull request by misaka36**
Release notes:
    https://sourceware.org/pipermail/binutils/2023-January/125671.html

Also remove old sha256sum.

Signed-off-by: Linhui Liu <liulinhui36@gmail.com>

### CODE REVIEW

File: toolchain/binutils/patches/2.40/039-LoongArch-ld-Fix-relocation-error-of-pcrel.patch [hunk 1]
```diff
@@ -0,0 +1,128 @@
+From 509a2ec6ad3ea7eb3f4cf59538cf636a2126e4c3 Mon Sep 17 00:00:00 2001
+From: liuzhensong <liuzhensong@loongson.cn>
+Date: Fri, 2 Sep 2022 16:29:14 +0800
+Subject: [PATCH 039/160] LoongArch:ld: Fix relocation error of pcrel.
```
Comment (hauke) [inline toolchain/binutils/patches/2.40/039-LoongArch-ld-Fix-relocation-error-of-pcrel.patch:4]: This is a backport from the 2.39 stable branch. It should be included in binutils 3.40.
I am not sure if it was really added to the 2.40 branch, but I think you should remove this patch.

Comment (lovehackintosh) [reply toolchain/binutils/patches/2.40/039-LoongArch-ld-Fix-relocation-error-of-pcrel.patch:4]: This patch didn’t add to the 2.40 branch. But I will remove this.

Comment (hauke) [reply toolchain/binutils/patches/2.40/039-LoongArch-ld-Fix-relocation-error-of-pcrel.patch]: It is strange that this patch is not in the 2.40 branch.
Probably there is a reason they only added it to the 2.39 branch. Maybe they have a better fix in 2.40.

**Comment by dhewg** [comment]
`Tested-by: Andre Heider <a.heider@gmail.com> # ipq40xx/generic`

Is there a reason to not use this version as new default?:
```
diff --git a/toolchain/binutils/Config.in b/toolchain/binutils/Config.in
index dbcaf4afce..e53384a006 100644
--- a/toolchain/binutils/Config.in
+++ b/toolchain/binutils/Config.in
@@ -2,7 +2,7 @@
 
 choice
    prompt "Binutils Version" if TOOLCHAINOPTS
-   default BINUTILS_USE_VERSION_2_39
+   default BINUTILS_USE_VERSION_2_40
    help
      Select the version of binutils you wish to use.
 
diff --git a/toolchain/binutils/Config.version b/toolchain/binutils/Config.version
index 479a8459fe..c9f8584480 100644
--- a/toolchain/binutils/Config.version
+++ b/toolchain/binutils/Config.version
@@ -6,10 +6,10 @@ config BINUTILS_VERSION_2_38
    bool
 
 config BINUTILS_VERSION_2_39
-   default y if !TOOLCHAINOPTS
    bool
 
 config BINUTILS_VERSION_2_40
+   default y if !TOOLCHAINOPTS
    bool
 
 config BINUTILS_VERSION
```

**Comment by lovehackintosh** [comment]
@dhewg Thanks for your patch, but I think we need more tests to make sure it's ok.

**Comment by lovehackintosh** [comment]
Oh, I'm going to create a new branch and apply your patch for testing.

**Comment by lovehackintosh** [comment]
See this: https://github.com/lovehackintosh/openwrt/tree/binutils-2.40-test

**Comment by Ansuel** [comment]
> @dhewg Thanks for your patch, but I think we need more tests to make sure it's ok.

usually we switch to new version after a while since these tool can contain very quiet bug so we wait a bit... (see gcc11 to gcc12 change)

**Comment by dhewg** [comment]
Sounds good to me!

Look that CI found: https://github.com/openwrt/openwrt/blob/master/include/trusted-firmware-a.mk#L84
All build failures are because of that afaict

**Comment by Ansuel** [comment]
@dhewg for that kind of thing... we should invert the condition... similar approach we follow for kernel bump and kernel modules

File: include/trusted-firmware-a.mk [hunk 2]
```diff
@@ -81,7 +81,7 @@ define Build/Compile/Trusted-Firmware-A
 		$(if $(DTC),DTC="$(DTC)") \
 		PLAT=$(PLAT) \
 		BUILD_STRING="OpenWrt v$(PKG_VERSION)-$(PKG_RELEASE) ($(VARIANT))" \
-		$(if $(CONFIG_BINUTILS_VERSION_2_39),LDFLAGS="-no-warn-rwx-segments") \
+		$(if $(CONFIG_BINUTILS_VERSION_2_39),$(CONFIG_BINUTILS_VERSION_2_40),LDFLAGS="-no-warn-rwx-segments") \
```
Comment (dhewg) [inline include/trusted-firmware-a.mk:84]: I think Ansuel meant to negate that check so we don't have to extend it again for newer versions in the future.
Untested, but something like `$(if $(CONFIG_BINUTILS_VERSION_2_37)$(CONFIG_BINUTILS_VERSION_2_38),,LDFLAGS="-no-warn-rwx-segments")`

Comment (lovehackintosh) [reply include/trusted-firmware-a.mk:84]: It looks fine. I will test it with the `binutils-2.40-test` branch

Comment (lovehackintosh) [reply include/trusted-firmware-a.mk]: Changed that. Everything worked for me, thanks.

File: include/trusted-firmware-a.mk [hunk 3]
```diff
@@ -81,7 +81,7 @@ define Build/Compile/Trusted-Firmware-A
 		$(if $(DTC),DTC="$(DTC)") \
 		PLAT=$(PLAT) \
 		BUILD_STRING="OpenWrt v$(PKG_VERSION)-$(PKG_RELEASE) ($(VARIANT))" \
-		$(if $(CONFIG_BINUTILS_VERSION_2_39),LDFLAGS="-no-warn-rwx-segments") \
+		$(if $(CONFIG_BINUTILS_VERSION_2_37)$(CONFIG_BINUTILS_VERSION_2_38),,LDFLAGS="-no-warn-rwx-segments") \
```
Comment (hauke) [inline include/trusted-firmware-a.mk:84]: Please add line breaks after 72 characters in the commit message.
Please move this commit before adding version 2.40.

Comment (lovehackintosh) [reply include/trusted-firmware-a.mk:84]: Ok, thanks.

**Pull request merged**